DOCS-#2954: Add docs for `modin.engines.base.frame` module (#2955)
modin-project/modin: docs/developer/flow/engines/base/frame/partition_manager.rst

BaseFrameManager
""""""""""""""""

The class is base for any partition manager class of ``pandas`` backend and serves as
intermediate level between ``pandas`` base frame and conforming :doc:`partition <partition>` class.
The class is responsible for partitions manipulation and applying a function to individual partitions:
block partitions, row partitions or column partitions, i.e. the class can form axis partitions from
block partitions to apply a function if an operation requires access to an entire column or row.
The class translates frame API into partition API and also can have some preprocessing operations
depending on the partition type for improving performance (for example,
:meth:`~modin.engines.base.frame.partition_manager.BaseFrameManager.preprocess_func`).

Main task of partition manager is to keep knowledge of how partitions are stored and managed
internal to itself, so surrounding code could use it via lean enough API without worrying about
implementation details.

Partition manager can apply user-passed (arbitrary) function in different modes:

> * along both axes piecewise:

Reviewer: ```suggestion
* block-wise (apply a function to individual block partitions):
```

  * optinally accepting partition indices along each axis
  * optionally accepting an item to be split so parts of it would be sent to each partition

> * along one axis piecewise

Reviewer: ```suggestion

```

> * along one full axis (when user function needs information about the whole axis)

Reviewer: ```suggestion
* along a full axis (apply a function to an entire column or row made up of block partitions when user function needs information about the whole axis)
```
